megatron/rl/sequence_packing_utils.py

```diff
@@ -0,0 +1,1147 @@
+        )
+        packer = SequencePacker(
+            bin_size=bin_size,
+            pad_token=tokenizer.pad,
+            max_sequences_per_bin=max_sequences_per_bin,
+        )
+
+        # Pack sequences with generation masks
+        (
+            packed_trajs,
+            packed_position_ids,
+            packed_attention_mask,
+            packed_loss_mask,
+            packing_info,
+        ) = packer.pack_sequences(trajs, generation_masks)
+
+        # Distribute packed bins across the data parallel ranks
+        (
+            packed_trajs,
+            packed_position_ids,
+            packed_attention_mask,
+            packed_loss_mask,
+            packing_info,
+        ) = distribute_packed_bins(
+            packed_trajs,
+            packed_position_ids,
+            packed_attention_mask,
+            packed_loss_mask,
+            packing_info,
+        )
+
+    # Create bin_advantages list
+    bin_advantages = []
+    for seq_indices in packing_info.bin_seq_indices:
+        if seq_indices:
+            bin_advantages.append(global_advantages[seq_indices])
+        else:
+            bin_advantages.append(
+                torch.tensor([], dtype=global_advantages.dtype, device=global_advantages.device)
+            )
+
+    # Pre-compute all PackedSeqParams for all bins ONCE to avoid repeated
+    # tensor allocations that cause CUDA memory fragmentation and periodic spikes
+    # Create a temporary packing context to pass to create_packed_seq_params
+    cached_packed_seq_params = [
+        create_packed_seq_params_for_bin(
+                packing_info=packing_info,
+                bin_idx=bin_idx,
+                bin_size=bin_size,
+                device=packed_trajs.device,
+            ) for bin_idx in range(len(packed_trajs))
+    ]
+
+    # Create the final PackingContext
+    packing_context = PackingContext(
+        bin_size=bin_size,
+        packer=packer,
+        packing_info=packing_info,
+        original_generation_masks=generation_masks,
+        original_trajs=trajs,
+        packed_trajs=packed_trajs,
+        packed_position_ids=packed_position_ids,
+        packed_attention_mask=packed_attention_mask,
+        packed_loss_mask=packed_loss_mask,
+        original_inference_logprobs=inference_logprobs,
+        bin_advantages=bin_advantages,
+        cached_packed_seq_params=cached_packed_seq_params,
+    )
+
+    log_packing_efficiency(packing_context)
+
+    return packing_context
+
+
+def get_microbatch_dataloader(packing_context: PackingContext) -> Tuple[DataLoader, int]:
+    args = get_args()
+    num_bins_this_rank = len(packing_context.packed_trajs)
+    dp_world_size = mpu.get_data_parallel_world_size()
+
+    # Ratio of collected sequences to the global batch size
+    pct_of_sequences_per_batch = len(packing_context.packing_info.seq_lengths) / args.global_batch_size
+    pct_of_bins_per_batch = pct_of_sequences_per_batch
```

Comment:
this line is a bit weird, let's just use pct_of_sequences_per_batch in the math.ceil expr